add a scope for `if let` guard temporaries and bindings

dianne opened the pull request:
This fixes my concern with `if let` guard drop order, namely that the guard's bindings and temporaries were being dropped after their arm's pattern's bindings, instead of before (https://github.com/rust-lang/rust/pull/141295#issuecomment-2968975596). The guard's bindings and temporaries now live in a new scope, which extends until (but not past) the end of the arm, guaranteeing they're dropped before the arm's pattern's bindings.

This only introduces a new scope for match arms with guards. Perf results (https://github.com/rust-lang/rust/pull/143376#issuecomment-3034922617) seemed to indicate there wasn't a significant hit to introduce a new scope on all match arms, but guard patterns (rust-lang/rust#129967) will likely benefit from only adding new scopes when necessary (with some patterns requiring multiple nested scopes).

Tracking issue for `if_let_guard`: rust-lang/rust#51114

Tests are adapted from examples by @traviscross, @est31, and myself on rust-lang/rust#141295.

dianne committed:
add more tests for `if let` guard drop order

dianne committed:
add a scope for `if let` guard temporaries and bindings

This ensures `if let` guard temporaries and bindings are dropped before
the match arm's pattern's bindings.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

Nadrieril commented:
I approve of the idea! Can't speak of the impl, I'm not familiar with MIR scopes.

r? compiler

dianne commented on tests/ui/drop/if-let-guards.rs:
```diff
@@ -0,0 +1,87 @@
+//! Tests drop order for `if let` guard bindings and temporaries. This is for behavior specific to
+//! `match` expressions, whereas `tests/ui/drop/drop-order-comparisons.rs` compares `let` chains in
+//! guards to `let` chains in `if` expressions.
```
Added: [diff](https://github.com/rust-lang/rust/compare/f501cdd77864f37c7995ec1231f5903261dfa045..8e5580a163d0f13cfdda0fc498f8d688722ca0cc). I noticed a comment I wrote about the lifetime of a `match` scrutinee's could be misleading, so I've amended that as well. It's still vague, but I think where *exactly* the scrutinee is dropped in that example is [technically Edition-dependent](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-tail-expr-scope.html) (just not in a way that affects these tests).

dianne committed:
only introduce a guard scope for arms with guards

dianne commented:
Rebased, cleaned up a little, tweaked comments, and made it so only match arms with guards introduce a new scope. I think this is ready for review now. r? @matthewjasper or reassign

matthewjasper commented:
@bors r+

bors commented:
:pushpin: Commit 0bdaef5b63fe2d557483e72732e463b86ffb041b has been approved by `matthewjasper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=matthewjasper 0bdaef5b63fe2d557483e72732e463b86ffb041b -->
<!-- homu: {"type":"Approved","sha":"0bdaef5b63fe2d557483e72732e463b86ffb041b","approver":"matthewjasper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 0bdaef5b63fe2d557483e72732e463b86ffb041b with merge a9447e2474503a09b7fb002a601d6a2f72522f9c...
<!-- homu: {"type":"BuildStarted","head_sha":"0bdaef5b63fe2d557483e72732e463b86ffb041b","merge_sha":"a9447e2474503a09b7fb002a601d6a2f72522f9c"} -->

Kobzol commented:
@bors retry
